Add Starpass Information to the overlay-view

reflexdd opened the pull request:
Starpass will be shown on the overlay with the same behaviours as on the scorebard-view. This works with Jammer names entered or not. Color of the SP is linked to the teamcolor.

reflexdd committed:
Add Starpass Information to the overlay-view

Starpass will be shown on the overlay with the same behaviours as on the scorebard-view. This works with Jammer names entered or not. Color of the SP is linked to the teamcolor.

reflexdd commented:
![SP-Scoreboard-Overlay-Status-0](https://user-images.githubusercontent.com/58275718/74089948-99961d80-4aa6-11ea-9e63-5d455963e72a.jpeg)
![SP-Scoreboard-Overlay-Status-1](https://user-images.githubusercontent.com/58275718/74089954-a9156680-4aa6-11ea-8941-60cd36d3b0aa.jpeg)
![SP-Scoreboard-Overlay-Status-4](https://user-images.githubusercontent.com/58275718/74093120-fa851c00-4acd-11ea-9d3a-8bca3c6cadb4.jpeg)
![SP-Scoreboard-Overlay-Status-2](https://user-images.githubusercontent.com/58275718/74089976-c77b6200-4aa6-11ea-9880-1086d3f0abec.jpeg)
![SP-Scoreboard-Overlay-Status-3](https://user-images.githubusercontent.com/58275718/74089978-c9452580-4aa6-11ea-8f12-2406a0dd3bcc.jpeg)

brian-brazil commented:
This looks pretty good.

I could see that being a little confusing when there's both a SP and a skater name. Could the separation be clearer without getting over complicated?

reflexdd commented:
I will see if i can make the separation a bit clearer.
Meanwhile I was playing a bit around with a re-styling of the overlay and found the solution quite nice having SP and Lead status on the left side.
![SP-Scoreboard-Overlay-Status_V2-1](https://user-images.githubusercontent.com/58275718/74201882-3acdd100-4c6b-11ea-902a-e4e1f536f3cf.jpeg)

brian-brazil commented:
That would work if that element is a little wider as you have there (and gives us more space to make the TTO/OR indicators more readable in future). The square corners don't look the best though.

reflexdd commented:
Wich element would you make wider? The SP/Lead Box or the Team-Name Section?
We can try make the left boxes a little wider. We then have to think about what to show if there is no color assigned. Maybe a small grey separator to indicate that there could be a color assigned.  I will see if i find a good solution to give TO/OR more space.

reflexdd commented:
I made a little sketch on how I can imagine to have TO/OR more visible
![scribble-scoreboard](https://user-images.githubusercontent.com/58275718/74268924-f5f87780-4d08-11ea-9c5f-52fb558c42eb.jpeg)

brian-brazil commented:
> Which element would you make wider? The SP/Lead Box or the Team-Name Section?

The SP/Lead box. The current checked in version is a bit narrow for what you're proposing.

> I will see if i find a good solution to give TO/OR more space.

Let's not have too much scope creep here. We can discuss this in a future PR.

reflexdd commented:
Ok coming back to the original change. I would see two possibilities to make it more distinguishable. 
The first would be to add a grey border to the SP Box when skaternames are set. That's also the one I would probably prefer.
![Screenshot_2020-02-15-Screenshot-3](https://user-images.githubusercontent.com/58275718/74589087-951bb880-5002-11ea-930b-a132a096a26c.jpeg)

The other one would be to use the same silver background for these boxes.
![Screenshot_2020-02-15-Screenshot-1](https://user-images.githubusercontent.com/58275718/74588861-6ef51900-5000-11ea-9fe2-0e5566a3064e.jpeg)
![Screenshot_2020-02-15-Screenshot-2](https://user-images.githubusercontent.com/58275718/74588863-70264600-5000-11ea-8a93-6f385941eb0e.jpeg)

reflexdd committed:
Update index.css

brian-brazil commented:
I think putting it in the left box like you suggested before looks best.

reflexdd commented:
Sure. Would you leave the Lead information where it is now or put it also in that box, like I suggested. I think having it on the left side makes it clearer to read.

brian-brazil commented:
Like you had in https://github.com/rollerderby/scoreboard/pull/436#issuecomment-584421068, but keeping the rounded corners.

reflexdd committed:
Overlay change for displaying Starpasses and Lead-Status

reflexdd commented:
Changes are now made to display Lead and Star Pass in the left Box.
![Screenshot_2020-02-16-Screenshot-1](https://user-images.githubusercontent.com/58275718/74605517-09209400-50c9-11ea-87ac-697906e315a3.jpeg)
![Screenshot_2020-02-16-Screenshot-2](https://user-images.githubusercontent.com/58275718/74605519-0a51c100-50c9-11ea-9f7d-d0c289be7fde.jpeg)
![Screenshot_2020-02-16-Screenshot-3](https://user-images.githubusercontent.com/58275718/74605520-0a51c100-50c9-11ea-9d9c-a9d1b09ac23e.jpeg)

brian-brazil commented on html/views/common.js:
```diff
@@ -20,6 +20,7 @@ function jammer(k, v) {
 	};
 	$(".Team" + id + " .Lead").toggleClass("HasLead", (leadJammer && !starPass));
 	$(".Team" + id).toggleClass("HasJammerName", (jn != ""));
+	$(".Team" + id).toggleClass("IsStarPass", (starPass != false));
```
Why not just `starPass` ?

reflexdd replied:
True, that was outdated as it resultetd from a previous version, wich I had.

brian-brazil commented on html/views/common.js:
```diff
@@ -9,7 +9,7 @@ function jammer(k, v) {
 
 	if (jammerName == null || jammerName == "") {
-		jammerName = leadJammer ? "Lead" : "";
+		jammerName = leadJammer ? "" : "";
```
Will this break the main scoreboard when there's a lead jammer we don't know the name of?

reflexdd replied:
You are right. I will put it back

reflexdd replied:
But it would be nice to not have the "Lead" name on the overlay.

reflexdd replied:
I'm not that expert in JS. Would there be a posibility to check if the function is called from the overlay wihtout duplicating the whole function and call that one out of the overlay?

brian-brazil replied:
This isn't a JS thing, it's CRG specific. This particular function is also kinda hairy, and has been the source of a string of bugs.

So I'd like to keep the logic together, so we don't inadvertently break the overlay when changing the scoreboard display.

I'd suggest a small wrapper function to pass in a bool. Though at this stage we probably should be looking at using WS.Register rather than continuing to abuse sbModify.

reflexdd replied:
I moved this to the overlay index.js and used WS.Register instead of sbModify to display Lead,Starpass and Skaternames

brian-brazil commented on html/views/common.js:
```diff
@@ -61,7 +62,7 @@ function smallDescriptionUpdate(k, v) {
 	var ret = '';
 
-	console.log(lc,tc,to,or,lcn,tcn);
+	// console.log(lc,tc,to,or,lcn,tcn);
```
You can delete this.

brian-brazil commented on html/views/overlay/index.js:
```diff
@@ -81,18 +81,32 @@ function initialize() {
 			}
-			if(v != null) style.addRule(rd, ns + ': ' + v);
+
+			for(var r=0; r<style.rules.length ; r++ ) {
```
This is getting messy.  Can we switch this to working like the PLT does?

reflexdd replied:
I replaced that now

reflexdd committed:
Update common.js

reflexdd committed:
Update index.js

reflexdd committed:
clean-up of Starpass class

reflexdd committed:
Update index.css

change of rounded corners on the right side of the number to rectangular.

brian-brazil commented on html/views/common.js:
```diff
@@ -20,6 +20,7 @@ function jammer(k, v) {
 	};
 	$(".Team" + id + " .Lead").toggleClass("HasLead", (leadJammer && !starPass));
 	$(".Team" + id).toggleClass("HasJammerName", (jn != ""));
+	$(".Team" + id).toggleClass("starPass", (starPass != false));
```
Did you forget to push?

reflexdd replied:
No, I changed
`$(".Team" + id).toggleClass("IsStarPass ", (starPass != false));` to
`$(".Team" + id).toggleClass("starPass", (starPass != false));`
and removed the redundant class IstStarPass in the css of the html.

brian-brazil replied:
You can still simplify to `starPass`

reflexdd replied:
Ok, got your point.

brian-brazil commented:
From a quick test this looks good and nothing seems to have been broken. The lead when there's no jammer name just needs resolving.

reflexdd committed:
removed redundant code

reflexdd committed:
Update common.js

reflexdd committed:
Update index.html

reflexdd committed:
Replacement of sbmodify for Jammer-Names

to remove the "Lead-String when no Jammer is selected changedthe way to display Lead, Starpass and Jammer/Pivot-Names

reflexdd committed:
Update index.js

brian-brazil commented on html/views/common.js:
```diff
@@ -20,6 +20,7 @@ function jammer(k, v) {
 	};
 	$(".Team" + id + " .Lead").toggleClass("HasLead", (leadJammer && !starPass));
 	$(".Team" + id).toggleClass("HasJammerName", (jn != ""));
+	$(".Team" + id).toggleClass("starPass", starPass);
```
HasStarPass for consistency

brian-brazil commented on html/views/overlay/index.css:
```diff
@@ -45,20 +45,22 @@ div 		{ white-space: nowrap; display: block; padding: 0; margin: 0;  overflow: h
+.Teambox .Indicator		{ width: 4%; height: 100%; font-size: 45%;  float: left; text-align: center;}
+.Teambox .Indicator .Lead		{ opacity: 0; height: 100%; }
+.Teambox .Indicator .HasLead	{ opacity: 1; }
+.TeamBox .starPass .Lead{ display: none; }
```
Isn't starPass also under Indicator?

Whitespace is also off here

reflexdd replied:
> Isn't starPass also under Indicator?

It's under Indicator, but the  .starpass Class is added when a Starpass is happening and will then hide the Lead div.

reflexdd replied:
> Whitespace is also off here

Wich whitespace do you mean here? The css property or the whitespace in the whole stylsheet?

brian-brazil replied:
> It's under Indicator, but the .starpass Class is added when a Starpass is happening and will then hide the Lead div.

My point is let's keep this consistent, and mention Indicator for both.

> Wich whitespace do you mean here?

`.Lead{`.

reflexdd replied:
Ok, i think i see what you mean. I will change it to 
`.TeamBox .Indicator .HasStarPass  { display: none; }`

brian-brazil commented on html/views/overlay/index.css:
```diff
@@ -45,20 +45,22 @@ div 		{ white-space: nowrap; display: block; padding: 0; margin: 0;  overflow: h
 .Team .HasLead		{ opacity: 1; color: red; }
 
-.Team .DotTimeouts	{ width: 3%; }
+.Team .DotTimeouts	{ width: 3%; padding-top: 2px}
```
semicolon

brian-brazil commented on html/views/overlay/index.js:
```diff
@@ -25,6 +25,46 @@ function initialize() {
 			'ScoreBoard.Period(*).Jam(*).TeamJam(*).Lead',
 			'ScoreBoard.Period(*).Jam(*).TeamJam(*).Lost'], function(k,v) { jamData(k,v); } );
 
+	WS.Register( ['ScoreBoard.Team(*).DisplayLead'], function(k,v) {
```
You need to register StarPass too, otherwise you can miss updates to starpass that don't affect DisplayLead

In general you must register everything you access.


This is also effectively re-writing jammerName, so that it's now implemented in two places which I'd like to avoid. I'd suggest hacking around what's there for now, and refactoring/reworking all that later.

reflexdd replied:
That's my impression too. it's getting messy having almost the same funcionality in two places.

reflexdd committed:
removed "Lead" as name displayed in OV

brian-brazil commented on html/views/overlay/index.js:
```diff
@@ -191,7 +151,26 @@ function initialize() {
 }
 
-
+function jammer_ov(k, v) {
```
This is still duplicating, try to share as much code as you sanely can with the existing function.

I very much look forward to cleaning all of this up, but that's a bigger effort.

reflexdd replied:
I tryed now to use the orginal jammer function by calling it from the jammer_ov function and handing over a bool. That seems to work without breaking the original function.

reflexdd committed:
Avoid duplicating jammer function

brian-brazil commented on html/views/common.js:
```diff
@@ -9,7 +9,7 @@ function jammer(k, v) {
 
 	if (jammerName == null || jammerName == "") {
-		jammerName = leadJammer ? "Lead" : "";
+		if(!ov)	jammerName = leadJammer ? "Lead" : "";
```
Please don't use if without braces, it's fragile.

frank-weinberg replied:
I think `jammerName = leadJammer && !ov ? "Lead" : ""; `  would be better.

brian-brazil replied:
If you're doing that, put parens around the condition for clarity.

brian-brazil replied:
Did you forget to push? I'm not seeing the change Speedy requested.

reflexdd replied:
I missed the comment. Will push it now.

brian-brazil commented on html/views/overlay/index.js:
```diff
@@ -165,7 +151,10 @@ function initialize() {
 
-
+function jammer_ov(k, v) {
+	var jn=	jammer(k,v,true);
```
`return jammer(k,v,true);`

brian-brazil commented:
Looking good, just some small style stuff.

reflexdd committed:
style changes

reflexdd committed:
Update index.css

fix of logo size and margins in Penaltyboard/Roster view to make it look equal

reflexdd committed:
Update common.js

brian-brazil commented:
Thanks!
